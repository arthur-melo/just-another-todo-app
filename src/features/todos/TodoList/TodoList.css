.todo_list__todos__li {
  opacity: 0;
  animation: fadeIn 0.3s 1;
  animation-fill-mode: forwards;
}

.todo_list__todos__li:last-child {
  margin-bottom: 0.5em;
}
