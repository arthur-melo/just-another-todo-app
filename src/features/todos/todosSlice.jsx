import { createSlice } from '@reduxjs/toolkit';
import { v1 } from 'uuid';

import initialState from '../initialState';

const todosSlice = createSlice({
  name: 'todos',
  initialState,
  reducers: {
    populate: (state, action) => {
      state.items = action.payload;
    },
    addItem: (state, action) => {
      const id = v1();
      const todoItem = {
        value: action.payload,
        id,
        completed: false,
      };

      state.items.push(todoItem);
    },
    cancelEditItem: state => {
      if (state.items.length) {
        state.editingItem = {};
      }
    },
    deleteItem: (state, action) => {
      const index = state.items.findIndex(item => item.id === action.payload);
      if (index !== -1) {
        state.items.splice(index, 1);
      }
    },
    editItem: (state, action) => {
      const index = state.items.findIndex(
        item => item.id === action.payload.id,
      );
      if (index !== -1) {
        state.items[index].value = action.payload.value;
      }
      state.editingItem = {};
    },
    itemCompletion: (state, action) => {
      const index = state.items.findIndex(
        item => item.id === action.payload.id,
      );
      if (index !== -1) {
        state.items[index].completed = !state.items[index].completed;
      }
    },
    markItemToEdit: (state, action) => {
      const index = state.items.findIndex(item => item.id === action.payload);

      if (index !== -1) {
        state.editingItem = state.items[index];
      }
    },
    reorderItem: (state, action) => {
      const [removed] = state.items.splice(action.payload.initial, 1);
      state.items.splice(action.payload.destination, 0, removed);
    },
  },
});

export const {
  populate,
  addItem,
  cancelEditItem,
  deleteItem,
  editItem,
  itemCompletion,
  markItemToEdit,
  reorderItem,
} = todosSlice.actions;

export const selectTodoItems = state => state.todos.items;
export const selectEditingItem = state => state.todos.editingItem;

export default todosSlice.reducer;
