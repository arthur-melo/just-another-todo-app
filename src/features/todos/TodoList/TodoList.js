import { useDispatch, useSelector } from 'react-redux';
import { DragDropContext, Droppable, Draggable } from '@hello-pangea/dnd';

import EditTodoItem from '../EditTodoItem/EditTodoItem';
import TodoItem from '../TodoItem/TodoItem';

import './TodoList.css';
import { selectTodoItems, selectEditingItem, reorderItem } from '../todosSlice';

const TodoList = () => {
  const dispatch = useDispatch();
  const todoItems = useSelector(selectTodoItems);
  const editingItem = useSelector(selectEditingItem);

  const onDragEnd = result => {
    // dropped outside the list
    if (!result.destination) {
      return;
    }

    dispatch(
      reorderItem({
        initial: result.source.index,
        destination: result.destination.index,
      }),
    );
  };

  return (
    <div className="container">
      <div className="row justify-content-center">
        <div className="col-12 col-md-9 col-lg-7 col-xl-6">
          <ul className="list-group">
            <DragDropContext onDragEnd={onDragEnd}>
              <Droppable droppableId="droppable">
                {provided => (
                  <div {...provided.droppableProps} ref={provided.innerRef}>
                    {todoItems.map((item, index) => (
                      <Draggable
                        key={item.id}
                        draggableId={item.id}
                        index={index}>
                        {provided => (
                          <div
                            ref={provided.innerRef}
                            data-testid="todo-list-item"
                            {...provided.draggableProps}
                            {...provided.dragHandleProps}
                            className="todo_list__todos__li">
                            {editingItem.id === item.id ? (
                              <EditTodoItem item={item} />
                            ) : (
                              <TodoItem item={item} />
                            )}
                          </div>
                        )}
                      </Draggable>
                    ))}
                    {provided.placeholder}
                  </div>
                )}
              </Droppable>
            </DragDropContext>
          </ul>
        </div>
      </div>
    </div>
  );
};

export default TodoList;
