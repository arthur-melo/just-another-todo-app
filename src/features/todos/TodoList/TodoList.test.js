import { screen, act } from '@testing-library/react';

import renderWithProviders from '../../../utils/testHelpers';
import initialState from '../../../features/initialState';
import TodoList from './TodoList';
import { addItem } from '../todosSlice';

describe('TodoList', () => {
  it('should render the TodoList component', () => {
    const view = renderWithProviders(<TodoList />);
    expect(view).toMatchSnapshot();
  });

  it('should render the TodoList component with preloaded items', () => {
    const testInitialState = {
      ...initialState,
      items: [
        {
          value: 'Test todo item',
          id: '0',
          completed: false,
        },
      ],
    };

    const view = renderWithProviders(<TodoList />, {
      preloadedState: {
        todos: testInitialState,
      },
    });

    expect(view).toMatchSnapshot();
  });

  it('should add an item to the todo items list', async () => {
    const testInitialState = {
      ...initialState,
      items: [
        {
          value: 'Test todo item',
          id: '0',
          completed: false,
        },
      ],
    };

    const { store } = renderWithProviders(<TodoList />, {
      preloadedState: {
        todos: testInitialState,
      },
    });

    let todoItems;
    todoItems = screen.getAllByTestId('todo-list-item');

    expect(todoItems).toHaveLength(1);

    act(() => store.dispatch(addItem('Second todo item')));

    todoItems = screen.getAllByTestId('todo-list-item');

    expect(todoItems).toHaveLength(2);
  });
});
