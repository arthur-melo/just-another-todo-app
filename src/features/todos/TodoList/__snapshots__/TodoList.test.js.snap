// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render the TodoList component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="row justify-content-center"
        >
          <div
            class="col-12 col-md-9 col-lg-7 col-xl-6"
          >
            <div
              class="todo_list__todos"
            >
              <ul
                class="list-group"
              >
                <div />
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="rfd-announcement-0"
      style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%);"
    />
    <div
      id="rfd-hidden-text-0-hidden-text-0"
      style="display: none;"
    >
      
  Press space bar to start a drag.
  When dragging you can use the arrow keys to move the item around and escape to cancel.
  Some screen readers may require you to be in focus mode or to use your pass through key

    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-12 col-md-9 col-lg-7 col-xl-6"
        >
          <div
            class="todo_list__todos"
          >
            <ul
              class="list-group"
            >
              <div />
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`TodoList should render the TodoList component with preloaded items 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="rfd-announcement-0"
      style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%);"
    />
    <div>
      <div
        class="container"
      >
        <div
          class="row justify-content-center"
        >
          <div
            class="col-12 col-md-9 col-lg-7 col-xl-6"
          >
            <div
              class="todo_list__todos"
            >
              <ul
                class="list-group"
              >
                <div>
                  <div
                    aria-describedby="rfd-hidden-text-1-hidden-text-2"
                    class="todo_list__todos__li"
                    data-rfd-drag-handle-context-id="1"
                    data-rfd-drag-handle-draggable-id="0"
                    data-rfd-draggable-context-id="1"
                    data-rfd-draggable-id="0"
                    data-testid="todo-list-item"
                    draggable="false"
                    role="button"
                    tabindex="0"
                  >
                    <li
                      class="todo_item__component list-group-item form-control"
                      data-testid="todo_item-listitem"
                    >
                      <div
                        class="text-truncate"
                        data-testid="todo_item-item-completion"
                      >
                        <div
                          class="todo_item__checkout d-inline-block"
                        >
                          <span>
                            <svg
                              aria-hidden="true"
                              class="svg-inline--fa fa-square fa-2x "
                              data-icon="square"
                              data-prefix="fas"
                              focusable="false"
                              role="img"
                              viewBox="0 0 448 512"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M0 96C0 60.7 28.7 32 64 32H384c35.3 0 64 28.7 64 64V416c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V96z"
                                fill="currentColor"
                              />
                            </svg>
                          </span>
                        </div>
                        <span
                          class="todo_item__text"
                        >
                          Test todo item
                        </span>
                      </div>
                    </li>
                  </div>
                </div>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="rfd-announcement-1"
      style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%);"
    />
    <div
      id="rfd-hidden-text-1-hidden-text-2"
      style="display: none;"
    >
      
  Press space bar to start a drag.
  When dragging you can use the arrow keys to move the item around and escape to cancel.
  Some screen readers may require you to be in focus mode or to use your pass through key

    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-12 col-md-9 col-lg-7 col-xl-6"
        >
          <div
            class="todo_list__todos"
          >
            <ul
              class="list-group"
            >
              <div>
                <div
                  aria-describedby="rfd-hidden-text-1-hidden-text-2"
                  class="todo_list__todos__li"
                  data-rfd-drag-handle-context-id="1"
                  data-rfd-drag-handle-draggable-id="0"
                  data-rfd-draggable-context-id="1"
                  data-rfd-draggable-id="0"
                  data-testid="todo-list-item"
                  draggable="false"
                  role="button"
                  tabindex="0"
                >
                  <li
                    class="todo_item__component list-group-item form-control"
                    data-testid="todo_item-listitem"
                  >
                    <div
                      class="text-truncate"
                      data-testid="todo_item-item-completion"
                    >
                      <div
                        class="todo_item__checkout d-inline-block"
                      >
                        <span>
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-square fa-2x "
                            data-icon="square"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 448 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M0 96C0 60.7 28.7 32 64 32H384c35.3 0 64 28.7 64 64V416c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V96z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </div>
                      <span
                        class="todo_item__text"
                      >
                        Test todo item
                      </span>
                    </div>
                  </li>
                </div>
              </div>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
